### PR TITLE
RavenDB-14318 Read snapshot to temp file before installing it

### PR DESCRIPTION
**Pull request by haludi**
https://issues.hibernatingrhinos.com/issue/RavenDB-14318

### CODE REVIEW

**Comment by grisha-kotler** [comment]
Can we have a test for this?

File: src/Raven.Server/Rachis/CandidateAmbassador.cs [hunk 1]
```diff
@@ -37,8 +37,8 @@ public string StatusMessage
         public volatile int ClusterCommandsVersion; // default is 0
         public string Tag => _tag;
 
-        private RemoteConnection _connection;
-        private RemoteConnection _publishedConnection;
+        private RemoteConnection.RemoteConnection _connection;
```
Comment (ayende) [inline src/Raven.Server/Rachis/CandidateAmbassador.cs]: why all these changes?

File: src/Raven.Server/Rachis/Follower.cs [hunk 2]
```diff
@@ -602,10 +605,28 @@ private void ReadAndCommitSnapshot(TransactionOperationContext context, InstallS
...
         }
 
-        private unsafe bool InstallSnapshot(TransactionOperationContext context, CancellationToken token)
+        private bool InstallSnapshot(TransactionOperationContext context, CancellationToken token)
         {
             var txw = context.Transaction.InnerTransaction;
-            var reader = _connection.CreateReader();
+
+            var fileName = $"snapshot.{Guid.NewGuid():N}";
+            var filePath = context.Environment.Options.DataPager.Options.TempPath.Combine(fileName);
```
Comment (ayende) [inline src/Raven.Server/Rachis/Follower.cs:614]: Can the temp path be null here?

File: src/Raven.Server/Rachis/Follower.cs [hunk 3]
```diff
@@ -602,10 +605,28 @@ private void ReadAndCommitSnapshot(TransactionOperationContext context, InstallS
...
+            using (var temp = new StreamsTempFile(filePath.FullPath, context.Environment))
+            using(var stream = temp.StartNewStream())
+            {
+                var remoteReader = _connection.CreateReaderToStream(stream);
+
+                if (ReadSnapshotFromFile(remoteReader, context, txw,false, token) == false)
```
Comment (ayende) [inline src/Raven.Server/Rachis/Follower.cs]: use named arg for the boolean

File: src/Raven.Server/Rachis/Follower.cs [hunk 4]
```diff
@@ -602,10 +605,28 @@ private void ReadAndCommitSnapshot(TransactionOperationContext context, InstallS
...
+                    return false;
+
+                stream.Seek(0, SeekOrigin.Begin);
+                var fileReader = new RemoteConnection.StreamReader(stream);
+                return ReadSnapshotFromFile(fileReader, context, txw, true, token);
+            }
+        }
+        
+        private unsafe bool ReadSnapshotFromFile(RemoteConnection.Reader reader, TransactionOperationContext context, Transaction txw, bool doWork, CancellationToken token)
```
Comment (ayende) [inline src/Raven.Server/Rachis/Follower.cs]: `doWork` name need to explain what it does better

File: src/Raven.Server/Rachis/RemoteConnection/RemoteConnection.cs [hunk 5]
```diff
@@ -12,7 +12,7 @@
 using Sparrow.Logging;
 using Sparrow.Server.Utils;
 
-namespace Raven.Server.Rachis
+namespace Raven.Server.Rachis.RemoteConnection
```
Comment (ayende) [inline src/Raven.Server/Rachis/RemoteConnection/RemoteConnection.cs]: What is this change necessary?

File: src/Raven.Server/Rachis/RemoteConnection/Rider.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
+using System;
```
Comment (ayende) [inline src/Raven.Server/Rachis/RemoteConnection/Rider.cs]: `Rider.cs` ?

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Rachis/CandidateAmbassador.cs (hunk 1 above)
Comment (haludi) [reply src/Raven.Server/Rachis/CandidateAmbassador.cs]: I moved some of the code (the riders) in RemoteConnection file to another file and put them both in a directory called RemoteConnection

**Comment by haludi** [comment]
> Can we have a test for this?
@grisha-kotler 
Yes, on progress

File: src/Raven.Server/Rachis/Follower.cs (hunk 2 above)
Comment (haludi) [reply src/Raven.Server/Rachis/Follower.cs:614]: No

File: src/Raven.Server/Rachis/RemoteConnection/RemoteConnection.cs (hunk 5 above)
Comment (haludi) [reply src/Raven.Server/Rachis/RemoteConnection/RemoteConnection.cs]: It just for better organization. 
But I will think about how to make it cleaner

File: src/Raven.Server/Rachis/RemoteConnection/Rider.cs (hunk 6 above)
Comment (haludi) [reply src/Raven.Server/Rachis/RemoteConnection/Rider.cs]: Oops, should be "Reader"

File: src/Raven.Server/Rachis/RemoteConnection/Rider.cs [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+using System.Diagnostics;
+using System.IO;
+using Sparrow.Binary;
+
+namespace Raven.Server.Rachis.RemoteConnection
```
Comment (ppekrol) [inline src/Raven.Server/Rachis/RemoteConnection/Rider.cs]: Maybe change to `Raven.Server.Rachis.Remote`. We will avoid `RemoteConnection.RemoteConnection` all over the place then.

File: src/Raven.Server/Rachis/RemoteConnection/Rider.cs [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+{
+    public class RemoteReaderToStream : RemoteReader
+        {
```
Comment (ppekrol) [inline src/Raven.Server/Rachis/RemoteConnection/Rider.cs]: Can you fix formatting?

File: src/Raven.Server/Rachis/Follower.cs (hunk 4 above)
Comment (ppekrol) [reply src/Raven.Server/Rachis/Follower.cs]: `dryRun` ?

File: src/Raven.Server/Rachis/Follower.cs (hunk 2 above)
Comment (ppekrol) [reply src/Raven.Server/Rachis/Follower.cs:614]: When this file will be deleted?

Comment (haludi) [reply src/Raven.Server/Rachis/Follower.cs:614]: In the `Dispose` of `StreamsTempFile`

File: src/Raven.Server/Rachis/Follower.cs (hunk 4 above)
Comment (ayende) [reply src/Raven.Server/Rachis/Follower.cs]: I like this

File: src/Raven.Server/Rachis/Follower.cs [hunk 9]
```diff
@@ -602,16 +607,36 @@ private void ReadAndCommitSnapshot(TransactionOperationContext context, InstallS
...
+                ReadSnapshot(fileReader, context, txw, dryRun: false, token);
+            }
+            
+            return true;
+        }
+        
+        private unsafe bool ReadSnapshot(Reader reader, TransactionOperationContext context, Transaction txw, bool dryRun, CancellationToken token)
+        {
+            var type = reader.ReadInt32();
+            if (type == -1)
```
Comment (ppekrol) [inline src/Raven.Server/Rachis/Follower.cs:636]: Is this correct? This was moved outside of the loop and switch statement doesn't have '-1' case

Comment (haludi) [reply src/Raven.Server/Rachis/Follower.cs:636]: Yes, -1 should be sent just in the beginning and in a case we don't need to send the snapshot
https://github.com/ravendb/ravendb/blob/6787a88d96c10c061e6b3d88525373c39a6e0873/src/Raven.Server/Rachis/FollowerAmbassador.cs#L510

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Server/Rachis/Follower.cs (hunk 9 above)
Comment (haludi) [reply src/Raven.Server/Rachis/Follower.cs:636]: I think in this way the meaning is more explicit and in a case, we will get "-1" in the middle we will get more describe error

File: src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs [hunk 10]
```diff
@@ -8,6 +8,7 @@
 using Raven.Client.Util;
 using Raven.Server.Json;
 using Raven.Server.Rachis;
+using Raven.Server.Rachis.Remote;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs:11]: remove

File: src/Raven.Server/Rachis/Candidate.cs [hunk 11]
```diff
@@ -6,6 +6,7 @@
 using Raven.Client.Exceptions;
 using Raven.Client.Http;
 using Raven.Client.ServerWide;
+using Raven.Server.Rachis.Remote;
```
Comment (karmeli87) [inline src/Raven.Server/Rachis/Candidate.cs:9]: remove

File: src/Raven.Server/Rachis/CandidateAmbassador.cs [hunk 12]
```diff
@@ -4,6 +4,7 @@
 using System.Threading;
 using Raven.Client.Http;
 using Raven.Client.ServerWide;
+using Raven.Server.Rachis.Remote;
```
Comment (karmeli87) [inline src/Raven.Server/Rachis/CandidateAmbassador.cs:7]: remove

File: src/Raven.Server/Rachis/Elector.cs [hunk 13]
```diff
@@ -1,6 +1,7 @@
 using System;
 using Raven.Client.Http;
 using Raven.Client.ServerWide;
+using Raven.Server.Rachis.Remote;
```
Comment (karmeli87) [inline src/Raven.Server/Rachis/Elector.cs:4]: remove

File: src/Raven.Server/Rachis/FollowerAmbassador.cs [hunk 14]
```diff
@@ -9,6 +9,7 @@
 using Raven.Client.Exceptions;
 using Raven.Client.Http;
 using Raven.Client.ServerWide;
+using Raven.Server.Rachis.Remote;
```
Comment (karmeli87) [inline src/Raven.Server/Rachis/FollowerAmbassador.cs:12]: remove

File: src/Raven.Server/Rachis/Leader.cs [hunk 15]
```diff
@@ -12,6 +12,7 @@
 using Raven.Server.Extensions;
 using Raven.Server.NotificationCenter.Notifications;
 using Raven.Server.NotificationCenter.Notifications.Details;
+using Raven.Server.Rachis.Remote;
```
Comment (karmeli87) [inline src/Raven.Server/Rachis/Leader.cs:15]: remove

File: src/Raven.Server/Rachis/RachisConsensus.cs [hunk 16]
```diff
@@ -31,8 +31,7 @@
 using Sparrow.Server;
 using Sparrow.Server.Utils;
 using System.Linq;
-using System.Reflection;
-using Voron.Data.BTrees;
+using Raven.Server.Rachis.Remote;
```
Comment (karmeli87) [inline src/Raven.Server/Rachis/RachisConsensus.cs:34]: remove

File: src/Raven.Server/Rachis/Remote/Reader.cs [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+        protected virtual int Read(int size)
+        {
+            if (Buffer.Length < size)
+                Buffer = new byte[Bits.PowerOf2(size)];
+            var totalRead = 0;
+            while (totalRead < size)
+            {
+                var read = InternalRead(totalRead, size - totalRead);
+                if (read == 0)
+                    break;
```
Comment (karmeli87) [inline src/Raven.Server/Rachis/Remote/Reader.cs]: Why `break` and not `throw new EndOfStreamException();`

File: src/Raven.Server/Rachis/Remote/RemoteConnection.cs [hunk 18]
```diff
@@ -270,66 +269,40 @@ public void Send(JsonOperationContext context, Exception e)
...
-                    var size = Math.Min(count, _buffer.Valid - _buffer.Used);
-                    fixed (byte* pBuffer = buffer)
-                    {
-                        Memory.Copy(pBuffer + offset, _buffer.Pointer + _buffer.Used, size);
-                        _buffer.Used += size;
-                        return size;
-                    }
-                }
+                var read = ReadFromBuffer(buffer, offset, count);
+                if (read != -1)
```
Comment (karmeli87) [inline src/Raven.Server/Rachis/Remote/RemoteConnection.cs]: Why is this change? It wasn't before, right?

Comment (haludi) [reply src/Raven.Server/Rachis/Remote/RemoteConnection.cs]: I just refactored the read from buffer code to a dedicated function

**Review by grisha-kotler — APPROVED** [review]

File: src/Raven.Server/Rachis/Remote/RemoteConnection.cs (hunk 18 above)
Comment (karmeli87) [reply src/Raven.Server/Rachis/Remote/RemoteConnection.cs]: okay, now I understand... I think a cleaner way will be something like this:
```
if (_buffer.Used < _buffer.Valid)
    return ReadFromBuffer(buffer, offset, count);

return _stream.Read(buffer, offset, count);
```

File: src/Raven.Server/Rachis/Follower.cs [hunk 19]
```diff
@@ -16,8 +19,10 @@
 using Sparrow.Utils;
 using Voron;
 using Voron.Data;
+using Voron.Data.BTrees;
 using Voron.Data.Tables;
 using Voron.Impl;
+using StreamReader = Raven.Server.Rachis.Remote.StreamReader;
```
Comment (ayende) [inline src/Raven.Server/Rachis/Follower.cs]: This is likely to be confusing WRT the `System.IO.StreamReader`, better to give a more specific name

File: src/Raven.Server/Rachis/Follower.cs [hunk 20]
```diff
@@ -602,16 +607,36 @@ private void ReadAndCommitSnapshot(TransactionOperationContext context, InstallS
...
+            var filePath = context.Environment.Options.DataPager.Options.TempPath.Combine(fileName);
+            using (var temp = new StreamsTempFile(filePath.FullPath, context.Environment))
+            using(var stream = temp.StartNewStream())
+            {
+                var remoteReader = _connection.CreateReaderToStream(stream);
+                if (ReadSnapshot(remoteReader, context, txw, dryRun:true, token) == false)
+                    return false;
+
+                stream.Seek(0, SeekOrigin.Begin);
+                var fileReader = new StreamReader(stream);
```
Comment (ayende) [inline src/Raven.Server/Rachis/Follower.cs]: `using` ?

File: src/Raven.Server/Rachis/Remote/Reader.cs [hunk 21]
```diff
@@ -0,0 +1,80 @@
+using System;
```
Comment (ayende) [inline src/Raven.Server/Rachis/Remote/Reader.cs]: Need to change the file name

File: src/Raven.Server/Rachis/Remote/Reader.cs [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            _parent = parent;
+        }
+
+        protected override int InternalRead(int offset, int count) => _parent.Read(Buffer, offset, count);
+    }
+    
+    public abstract class Reader
+    {
+        public byte[] Buffer { get; private set;} = new byte[1024];
```
Comment (ayende) [inline src/Raven.Server/Rachis/Remote/Reader.cs]: Shouldn't we use `ArrayPool` here?

File: src/Raven.Server/Rachis/Remote/Reader.cs [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+        public long ReadInt64()
+        {
+            ReadExactly(sizeof(long));
+            return BitConverter.ToInt64(Buffer, 0);
+        }
+
+        public virtual void ReadExactly(int size)
+        {
+            if (Buffer.Length < size)
+                Buffer = new byte[Bits.PowerOf2(size)];
```
Comment (ayende) [inline src/Raven.Server/Rachis/Remote/Reader.cs]: `ArrayPool`

File: src/Raven.Server/Rachis/Remote/Reader.cs [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+
+        public RemoteReader(RemoteConnection parent)
+        {
+            _parent = parent;
+        }
+
+        protected override int InternalRead(int offset, int count) => _parent.Read(Buffer, offset, count);
+    }
+    
+    public abstract class Reader
```
Comment (ayende) [inline src/Raven.Server/Rachis/Remote/Reader.cs]: Need a better name

**Review by ayende — APPROVED** [review]

**Review by karmeli87 — APPROVED** [review]
